walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,66 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Set;
+import java.util.HashSet;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Arrays;
+
+import com.google.sps.Event;
+import com.google.sps.TimeRange;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    Set<String> allAttendees = new HashSet<>(mandatoryAttendees);
+    allAttendees.addAll(optionalAttendees);
+    
+    Collection<TimeRange> ranges = getTimeRanges(events, allAttendees, request.getDuration());
+    if (ranges.size() > 0) {
+      return ranges;
+    }
+    if (mandatoryAttendees.size() > 0) {
+      return getTimeRanges(events, mandatoryAttendees, request.getDuration());
+    }
+    return Arrays.asList();
+  }
+
+  public Collection<TimeRange> getTimeRanges(Collection<Event> events, Collection<String> attendees, long duration) {
+    Comparator<Event> comparator = new Comparator<Event>() {
+      @Override
+      public int compare(final Event e1, final Event e2) {
+        return TimeRange.ORDER_BY_START.compare(e1.getWhen(), e2.getWhen());
+      }
+    };
+    boolean rejectedEvent = false;
+    List<Event> l = new ArrayList(events);
+    Collections.sort(l, comparator);
+    Collection<TimeRange> ranges = new ArrayList<TimeRange>();
+    int start = TimeRange.START_OF_DAY;
+    for (Event e : l) {
+      Set<String> eventAttendees = e.getAttendees();
+      Set<String> copy = new HashSet<>(eventAttendees);
```

Comment:
is there a more efficient way to accomplish this (new HashSet<>(...).retainAll(...))? for simple tricks like this, your first thought should generally be, has someone else already made a way to make this easier for me? if so, is it appropriate to use that instead?

Reply:
Hmmm.. I'm not sure. I need a set intersection operator to determine if the members of the meeting and the members of the event overlap. Unfortunately, Java's `retainAll` is mutable, so I needed to do some copying-nonsense.

Is there some method I'm missing? I believe I read the `HashSet` docs, and I couldn't find any mention of intersection.

Reply:
You're correct in that what you're currently doing is intersection, and your current code is doing intersection quite reasonably. Is intersection actually what your current code requires however? Your code does not actually use the results of the intersection, it merely tests whether the intersection is empty or not, ie if the sets are disjoint or not. This means that you're doing some extra work (not a huge deal here), but also that there might be a more effective method, such as [Collections.disjoint()](https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#disjoint-java.util.Collection-java.util.Collection-l).

How I arrived at this (I don't have the entire Java collections lib memorized sadly): the size() comparison to 0 is immediately suspicious to me. Collections usually provide an isEmpty() method which is always preferable to size() as it may be more efficient (such as for linked lists). Once I established that we really only cared about isEmpty() and not the size, then it becomes clear that you don't need the whole intersection, you only care about the disjunction. At that point, googling something like "java set disjunction" lead me to that method.

Reply:
I could also just do this with a simple for loop.